libraries/AP_Beacon/AP_Beacon_Nooploop.cpp

```diff
@@ -0,0 +1,238 @@
+        uint8_t b = _uart->read();
+        MsgType type = parse_byte(b);
+        if (type == MsgType::NODE_FRAME2) {
+            if (_anchor_pos_avail) {
+                parse_node_frame2();
+            } else {
+                request_setting();
+            }
+        } else if (type == MsgType::SETTING_FRAME0) {
+            parse_setting_frame0();
+        } 
+    }
+}
+
+void AP_Beacon_Nooploop::request_setting()
+{
+    if (_uart->txspace() >= 128) {
+        uint8_t buf[128] = {0x54, 0, 1}; 
+        buf[127] = 0x55; 
+        _uart->write(buf, 128); //send setting_frame0 to tag, tag will fill anchor position
+    }
+}
+
+// process one byte received on serial port
+// message is stored in _msgbuf
+AP_Beacon_Nooploop::MsgType AP_Beacon_Nooploop::parse_byte(uint8_t b)
+{
+    // process byte depending upon current state
+    switch (_state) {
+
+    case ParseState::HEADER:
+        if (b == NOOPLOOP_HEADER) {
+            _msgbuf[0] = b;
+            _msg_len = 1;
+            _crc_expected = b;
+            _state = ParseState::H55_FUNCTION_MARK;
+        } else if (b == NOOPLOOP_HEADER2) {
+            _msgbuf[0] = b;
+            _msg_len = 1;
+            _crc_expected = b;
+            _state = ParseState::H54_FUNCTION_MARK;
+        }
+        break;
+
+    case ParseState::H55_FUNCTION_MARK:
+        if (b == NOOPLOOP_FUNCTION_MARK_NODE_FRAME2) {
+            _msgbuf[1] = b;
+            _msg_len++;
+            _crc_expected += b;
+            _state = ParseState::LEN_L;
+        } else {
+            _state = ParseState::HEADER;
+        }
+        break;
+
+    case ParseState::H54_FUNCTION_MARK:
+        if (b == NOOPLOOP_FUNCTION_MARK_SETTING_FRAME0) {
+            _msgbuf[1] = b;
+            _msg_len++;
+            _crc_expected += b;
+            _state = ParseState::SF0_PAYLOAD;
+        } else {
+            _state = ParseState::HEADER;
+        }
+        break;
+
+    case ParseState::LEN_L:
+        _msgbuf[2] = b;
+        _msg_len++;
+        _crc_expected += b;
+        _state = ParseState::LEN_H;
+        break;
+
+    case ParseState::LEN_H:
+        // extract and sanity check frame length
+        _frame_len = (uint16_t)b << 8 | _msgbuf[2];
+        if (_frame_len > NOOPLOOP_NODE_FRAME2_FRAMELEN_MAX) {
+            _state = ParseState::HEADER;
+        } else {
+            _msgbuf[3] = b;
+            _msg_len++;
+            _crc_expected += b;
+            _state = ParseState::NF2_PAYLOAD;
+        }
+        break;
+
+    case ParseState::NF2_PAYLOAD:
+        // add byte to buffer if there is room
+        if (_msg_len < NOOPLOOP_MSG_BUF_MAX) {
+            _msgbuf[_msg_len] = b;
+        }
+        _msg_len++;
+        if (_msg_len >= _frame_len) {
+            _state = ParseState::HEADER;
+            // check crc
+            if (b == _crc_expected) {
+                return MsgType::NODE_FRAME2;
+            } else {
+                return MsgType::INVALID;
+            }
+        } else {
```

Comment:
Don't tneed this else either.

Reply:
Yes, I agree. But I think code would be easier to read this way